### DISCUSSION TITLE
Provide an equivalent to "labels"

### DISCUSSION THREAD

**Discussion by LekoArts** [community/community#19] [Discussions] [answered]
# Why

At [Gatsby](https://github.com/gatsbyjs/gatsby) we've made heavy use of different labels for displaying the `status`, `topic` and `type` of an issue. Since we're a monorepo with nearly 100 packages we need those labels to differentiate the different types like questions, feature requests, bugs for the respective package.

In moving questions and feature requests over to Discussions I've already seen that a lack of overview and ability to filter things will be a problem in the future. Thus we're in need of an equivalent to "labels".

# How

Since Discussions isn't something completely new, there's enough previous art to draw inspiration from:

- Tags (like StackOverflow)
  - Could be a predefined list (like labels) or free-form
- Subcategories (like in any forum software or Zendesk)
  - So that the sections that people can browse are nested
- Allow prefixes to title which can be used for filtering (like in any forum software)

**Comment by LekoArts** [comment]
@becca Hi! Can you give an initial comment, e.g. if this is tracked internally or questions are still open/to be discussed. As I'm not sure what 👀 means :) Appreciate any feedback, thanks!

**Comment by becca** [comment]
We're currently investigating having a shared system of `labels` between Issues and Discussions. 

Regarding the other mentioned features:
1. Subcategories: I'm curious how you would expect these to function. Is this more or less another level of organization mostly for the end user? Are there any behaviours you would expect subcategories to have?
2. Tags: Are you imagining having a separate list of tags in Discussions or would the shared labels system with Issues address your needs? If 
3. Prefixes: How might these differ from tags or labels? If you could share more details on this with relation to the filtering experience specifically that would be great!

**Reply by LekoArts** [reply]
> We're currently investigating having a shared system of labels between Issues and Discussions.

That would be ideal/best solution I think! Keeps the organization of those coherent and when transferring an issue to a discussion they can be transferred with it.

> Subcategories: I'm curious how you would expect these to function. Is this more or less another level of organization mostly for the end user? Are there any behaviours you would expect subcategories to have?

Yeah, more or less an organizational thing:

![image](https://user-images.githubusercontent.com/16143594/99043929-3f733d00-258f-11eb-8f82-6548a1822ac9.png)

It couldn't be a replacement for tags though as e.g. we have 100 of them and one shouldn't need to scroll through 100 subcategories 😅

**Are there any behaviours you would expect subcategories to have?** Subcategories could inherit the settings of the parent by default, e.g. when parent has setting that answered can be marked as answered the child categories would automatically also have that (but could be overriden)

> Tags: Are you imagining having a separate list of tags in Discussions or would the shared labels system with Issues address your needs?

As already mentioned above we'd be really happy about a shared system!

> Prefixes: How might these differ from tags or labels? If you could share more details on this with relation to the filtering experience specifically that would be great!

In the past I experienced prefixes like this (in forum software like https://invisioncommunity.com/4guides/how-to-use-ips-community-suite/content-discovery/content-tags-and-prefixes-r72/):

![image](https://user-images.githubusercontent.com/16143594/99044232-bc061b80-258f-11eb-88f1-7e3e11bf6f4b.png)

Prefix is definitely a niche thing though, I don't expect many people to know that UX pattern, especially compared to labels/tags.
The nice thing about them is that they allow the custom setting of boolean flags. What I mean with that is:

In the other discussion we talked about having a third state like "resolved" in addition to unanswered/answered. With prefixes you can add such toggles individually to each category. Let's say we have a category for feature requests, then one could add two prefixes called unresolved/resolved. By default unresolved gets applied and maintainers could change the prefix to resolved -- or it happens automatically.

That's at least behavior in forum softwares. But as I said, labels can solve most things and are probably more common.

**Reply by becca** [reply]
Thank you so much for expanding on your request. This is extremely helpful and I will keep you posted once we have labels available in Discussions. It's on our roadmap as we know this is a critical feature for maintainers to more thoroughly organize this space. 

For the "states" you mentioned- would you expect/need those as labels or perhaps believe those to be more of a feature that's available at the Category level (Q&A categories can have answerable/resolved states, for example)?

**Reply by LekoArts** [reply]
> For the "states" you mentioned- would you expect/need those as labels or perhaps believe those to be more of a feature that's available at the Category level (Q&A categories can have answerable/resolved states, for example)?

I'd say labels could serve the need of states, but having states like "Q&A categories can have answerable/resolved states, for example" would be complimentary for me. So it would certainly be a nice to have but labels are more important.

**Reply by becca** [reply]
Noted! Thank you for that additional context regarding the prioritization of labels vs additional states at the category level. 🙇

**Comment by gotom22** [comment]
Would like second this request for labels (and or similar organizing features)
_(but reaction button does not display correctly for me in chrome on mac. Thumbs up is outside of browser window. not displayed, clickable. - bug?)_

I would hope to have for Discussions similar sorting features like for Issues: by author, by label, most recent, and maybe by Topic. A topic-label could be used in a drop down to select filter by "main topics" - these would probably have to be set by an admin.

**Comment by Aviad-Herman** [comment]
Upvote

**Comment by becca** [comment]
We are aware of this feedback and are tracking the request internally. We'll keep you posted on the plan for this coming to Discussions!

**Reply by brandonkelly** [reply]
Well this is promising: https://twitter.com/natfriedman/status/1379847199639564288?s=21

**Comment by brandonkelly** [comment]
We (@craftcms) have been looking into migrating feature request from Issues over to Discussions, but the lack of labels has us reconsidering it. Our main repo, [craftcms/cms](https://github.com/craftcms/cms) has over 50 labels, most of which are for specific features of the product, and we rely on our ability to filter ideas by feature.

Subcategories wouldn’t work for us because there are cases where issues have 2+ labels, such as [ux + authoring](https://github.com/craftcms/cms/issues?q=is%3Aopen+is%3Aissue+label%3A%22ux+%3Asmile%3A%22+label%3A%22authoring+%3Awriting_hand%3A%22) or [ux + system administration](https://github.com/craftcms/cms/issues?q=is%3Aopen+is%3Aissue+label%3A%22ux+%3Asmile%3A%22+label%3A%22system+administration+%3Abriefcase%3A%22).

Ideally discussions would be able to use the same list of labels that issues (and pull requests) use.

We also wouldn’t want users to have to choose which labels to add to their discussions; they should just have to select from the discussion categories, and get to writing. Let us be the ones to pick the appropriate labels, as we do with issues.

**Comment by andrewleader** [comment]
In the same boat with Microsoft Project Reunion, we have about 7 feature areas and need to easily be able to view "Ideas"/etc per feature area (like Notifications, WinUI, Packaging, etc).

**Comment by evi-liu** [comment]
## Labels Private Beta Available 🚀 

Thank you for all of the feedback on labels! I'm excited to announce that we have a private beta available, where you can use your repository labels on Discussions as well. If you'd like to be included in the beta, please share the repository you'd like to be flagged in below. For now, this private beta is limited to the first 15 repositories that respond. 

![image](https://user-images.githubusercontent.com/8809849/116739245-4b6d8280-a9a8-11eb-85d7-d394778fdcda.png)

**Reply by brandonkelly** [reply]
[craftcms/cms](https://github.com/craftcms/cms) please!

**Reply by bnb** [reply]
Oh my god yes. nodejs/node would be nice ❤️

**Reply by andrewleader** [reply]
WAHOO!! Add [microsoft/ProjectReunion](https://github.com/microsoft/projectreunion) please!

**Reply by KyleAMathews** [reply]
✋ gatsbyjs/gatsby please!

**Reply by lukeholder** [reply]
[craftcms/commerce](https://github.com/craftcms/commerce) please!

**Reply by evi-liu** [reply]
everyone in this thread has been flagged in and now has labels available! please let us know if you have any feedback (either positive or negative!)

**Reply by brandonkelly** [reply]
Thanks @evi-liu! ❤️ 

Looks like labels are not currently preserved when converting existing issues to discussions via the “Convert Issues” button.

**Reply by jopdeklein** [reply]
Thanks for adding our repo @evi-liu! 

I noticed a small visual issue with labels that contain emoji's at the end being cut off. This works fine for PR's and issues.
![Screenshot 2021-05-04 at 14 03 37](https://user-images.githubusercontent.com/200943/117000701-c94fc780-ace1-11eb-8b6b-d1be0424e0a7.png)

**Reply by natowi** [reply]
https://github.com/alicevision/meshroom please :)

**Reply by Andre601** [reply]
https://github.com/botblock/JavaBotBlockAPI (No typo this time, unlike the other time... Did you get the notification btw?)
https://github.com/PlaceholderAPI/PlaceholderAPI
I hope it counts that I'm a outside contributor to the repository. I use the discussions there frequently to share info and this would be a useful addition.
@darbyjack may have extra info and/or can approve/deny my request here in the name of the org.

**Reply by evi-liu** [reply]
@natowi @Andre601 your repos have been flagged in!

@brandonkelly @jopdeklein thanks for reporting! we'll look into these :)

**Reply by zol** [reply]
@evi-liu we would love love love labels on [storybookjs/storybook](https://github.com/storybookjs/storybook/discussions) . Their existence is well timed as we have been considering moving back to issues because of a lack of labels/organization of discussions.

**Reply by evi-liu** [reply]
@zol enabled! glad we made it in the nick of time :)

**Reply by harshraj22** [reply]
@evi-liu I'd love to have it in my repo as well. [harshraj22/DeepLearning](https://github.com/harshraj22/DeepLearning). Please :smile:

**Reply by rijkvanzanten** [reply]
We'd love to test it out under [`directus/directus`](https://github.com/directus/directus) 😄 🐰

**Comment by evi-liu** [comment, marked as answer]
## Labels available as public beta! 🚀 

Thanks for all of the helpful feedback and private beta testing of labels. We've now rolled out labelling capabilities to all repositories with Discussions. Happy labelling and feel free to drop any feedback in this thread.

**Reply by laymonage** [reply]
Great! Will this be available through the API soon?

**Reply by evi-liu** [reply]
We're working on making it available for our GraphQL API soon!

**Reply by bnb** [reply]
@evi-liu is there an announcement I can share? <3

**Reply by evi-liu** [reply]
@bnb 

Here's the official changelog on GitHub blog: https://github.blog/changelog/2021-05-18-github-discussions-labels-and-announcements-category-format/

And I also made a more informal announcement here: https://github.com/github/feedback/discussions/3979

**Reply by aviflax** [reply]
What are the chances of getting this feature in team discussions? Or is that a different feature that I should be asking about somewhere else?

**Reply by evi-liu** [reply]
@aviflax Unfortunately, it's a different feature entirely. You can share feedback about it here https://support.github.com/contact/feedback

**Reply by aviflax** [reply]
Got it — thank you!
